### PR TITLE
Contrib/ltx 2.3

### PR DESCRIPTION
**Pull request by jimburtoft**
## Description
Adds LTX-2.3, a 22B parameter DiT (Diffusion Transformer) for joint audio-video generation, running on Trainium 2 via NxD Inference. The model generates synchronized video + audio from text prompts using bidirectional audio-video cross-attention with flow matching.
Both the 22B DiT backbone and Gemma 3 12B text encoder run on Neuron at TP=4, sharing 4 NeuronCores and executing sequentially. CPU handles VAE decode and optional latent upscaling. Uses the native [ltx-core](https://github.com/Lightricks/LTX-2) framework (not Diffusers).
Key results:
- Single forward pass cosine similarity: 0.9999 (video), 0.9999 (audio) vs CPU reference
- Neuron Gemma3 text encoding: 6.6s (83x faster than CPU fallback at ~162s)
- Warm denoising step: ~0.3s per step (22B DiT, 8 Euler steps)
- E2E output: 25 frames @ 384x512 video + stereo 48kHz audio
## Model Information
**Model Name:** LTX-2.3
**Model Architecture:** DiT (Diffusion Transformer) — 48 transformer blocks, 32 attention heads, 4096 video dim, 2048 audio dim. Bidirectional audio-video cross-attention, gated attention, QK-RMSNorm, split RoPE, flow matching.
**Purpose:** Joint text-to-video and text-to-audio generation with optional spatial/temporal latent upscaling.
## Checklist
Please ensure your PR includes the following items. Refer to the [contrib/CONTRIBUTING.md](../contrib/CONTRIBUTING.md) for detailed guidelines.
### Required Components
- [x] **Accuracy Test** (ex. `test/integration/test_model.py`)
  - At least one integration test that validates model accuracy
  - Uses logit validation or equivalent accuracy verification
  - Test can compile and run the model on Neuron
- [x] **README.md** with the following sections:
  - [x] **Usage Example**: Clear code example showing how to use the model
  - [x] **Compatibility Matrix**: Table showing tested Neuron SDK versions and instance types (Trn1/Trn2/Inf2)
  - [x] **Example Checkpoints**: Links to compatible model checkpoints (e.g., HuggingFace Hub)
  - [x] **Testing Instructions**: Command to run the test suite for the model
- [x] **Source Code** (`src/`)
  - Modeling code following NxD Inference patterns
  - Properly structured in the contrib folder hierarchy
### Optional Components
- [ ] **Unit Tests** (CPU or Neuron-based)
  - Tests for individual modeling components
  - Located in `test/unit/` directory
## Folder Structure
Confirm your contribution follows this structure:
/contrib/models/LTX-2.3/
  README.md
  /src
    init.py
    modeling_ltx23.py              # Core DiT backbone (TP sharding, DistributedRMSNorm, SDPA replacement)
    modeling_gemma3_encoder.py     # Custom Gemma3 encoder-only model (returns all 49 hidden states)
    pipeline.py                    # NeuronTransformerWrapper (CPU preprocessing, backbone routing)
    compile_transformer.py         # DiT backbone compilation script
    compile_gemma3.py              # Gemma3 encoder compilation script
    shard_gemma3_weights.py        # Pre-shard Gemma3 weights to per-rank files
    load_with_weights.py           # DiT backbone weight sharding and injection
    generate_ltx23.py              # E2E generation pipeline
  /test
    init.py
    /unit
      init.py
    /integration
      init.py
      test_model.py
## Testing
**How did you test this change?**
All testing performed on trn2.3xlarge (TP=4, LNC=2) with Neuron SDK 2.28, PyTorch 2.9, Deep Learning AMI Neuron (Ubuntu 24.04) 20260227.
Tests include:
1. **Integration tests** (4 tests): model loading, forward pass validity, accuracy vs CPU reference, latency measurement
2. **E2E generation**: Full pipeline with Neuron Gemma3 text encoding → Neuron DiT denoising (8 steps) → CPU VAE decode, producing 25 video frames @ 384x512 + stereo 48kHz audio
**Test Results:**
test/integration/test_model.py::test_model_loads           PASSED
test/integration/test_model.py::test_forward_pass_no_nan   PASSED
test/integration/test_model.py::test_accuracy_vs_cpu       PASSED
test/integration/test_model.py::test_performance_latency   PASSED
========================= 4 passed in 456.89s =========================
Accuracy validation (measured against unsharded BF16 CPU reference):
| Component | Metric | Value |
|-----------|--------|-------|
| Single forward pass (video) | Cosine similarity | 0.999947 |
| Single forward pass (audio) | Cosine similarity | 0.999867 |
| 8-step denoised latent (real text) | Cosine similarity | 0.972 |
## Compatibility
**Tested with:**
- **Neuron SDK Version(s):** 2.28
- **Instance Type(s):** trn2.3xlarge (TP=4, LNC=2)
- **PyTorch Version:** 2.9
- **Python Version:** 3.12
## Additional Information
- **Two Neuron models share NeuronCores**: Both the 22B DiT backbone and 12B Gemma3 text encoder are compiled for TP=4 and share the same 4 NeuronCores, executing sequentially (text encoding runs once, then denoising loop runs 8 steps).
- **Different compiler flags per model**: DiT uses `--auto-cast matmult` for throughput; Gemma3 uses `--auto-cast=none --enable-saturate-infinity --enable-mixed-precision-accumulation` for text encoder precision.
- **Cold start**: First two denoising steps are slow (~144s + ~177s) due to Neuron device initialization. Warm steps run at ~0.3s each.
- **CPU fallback available**: Gemma3 can run on CPU (~162s) without Neuron compilation, using `--gemma-path` without `--neuron-gemma`.
- **Stage 1 only**: This submission covers Stage 1 generation with optional latent upscaling. Stage 2 refinement denoising (recompiling at larger latent shapes) is planned as a follow-up.
- Uses native [ltx-core](https://github.com/Lightricks/LTX-2) framework, not Diffusers.
## Related Issues
N/A — first contribution for this model.
## vLLM Integration
- [ ] This model/feature is intended for use with vLLM
- [ ] Documentation includes vLLM registration instructions
---
**By submitting this PR, I confirm that:**
- [x] I have read and followed the [contributing guidelines](../contrib/CONTRIBUTING.md)
- [x] This is a community contribution and may have limited testing compared to officially-supported models
- [x] The code follows best practices and is well-documented
- [x] All required components listed above are included

### CODE REVIEW

**Review by sdeeptan-aws — COMMENTED** [review]
@jimburtoft There are multiple compilation and weight sharding scripts. Can we combine them?

**Review by jimburtoft — COMMENTED** [review]
Done — consolidated the scripts per your suggestion:

**Before** (7 scripts): `compile_transformer.py`, `compile_transformer_halfres.py`, `compile_gemma3.py`, `compile_vae_23.py`, `shard_backbone_weights.py`, `shard_gemma3_weights.py`, `compile_benchmark.py`

**After** (3 scripts):
- `compile.py` — unified compilation with subcommands: `transformer`, `encoder`, `vae`
- `shard_weights.py` — unified weight sharding with subcommands: `backbone`, `encoder`
- `compile_benchmark.py` — kept separate (trn2.48xlarge multi-chip benchmark only)

Also validated end-to-end on SDK 2.29 (DLAMI 20260410, neuronx-cc 2.24, torch-neuronx 2.9.0.2.13). Performance identical to 2.28: 0.3s/step warm denoising, 25 frames @ 384×512 + stereo audio. README updated with new commands and compatibility matrix.

File: contrib/models/LTX-2.3/src/load_with_weights.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+        f"mean={a_np.mean():.4f}, std={a_np.std():.4f}",
+        flush=True,
+    )
+
+    has_nan = np.isnan(v_np).any() or np.isnan(a_np).any()
+    has_inf = np.isinf(v_np).any() or np.isinf(a_np).any()
+    print(f"\n  NaN: {has_nan}, Inf: {has_inf}", flush=True)
+
+    # Save outputs
+    output_path = "/home/ubuntu/ltx23_neuron/test_outputs_real_weights.pt"
```
Comment (lutfanm-aws) [inline contrib/models/LTX-2.3/src/load_with_weights.py]: Same comment as above.

File: contrib/models/LTX-2.3/src/load_with_weights.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+import time
+import json
+import gc
+import torch
+import numpy as np
+
+os.environ["NEURON_FUSE_SOFTMAX"] = "1"
+os.environ["NEURON_CUSTOM_SILU"] = "1"
+os.environ["NEURON_RT_STOCHASTIC_ROUNDING_EN"] = "0"
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
```
Comment (lutfanm-aws) [inline contrib/models/LTX-2.3/src/load_with_weights.py]: Prefer os.environ.setdefault() so there is no chance of different scripts using different values

File: contrib/models/LTX-2.3/src/load_with_weights.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+MODEL_PATH = "/home/ubuntu/models/LTX-2.3/ltx-2.3-22b-distilled.safetensors"
+COMPILE_DIR = "/home/ubuntu/ltx23_neuron/compiler_workdir_tp4_lnc2_v2"
```
Comment (lutfanm-aws) [inline contrib/models/LTX-2.3/src/load_with_weights.py]: Hardcoded paths, please remove

**Review by lutfanm-aws — APPROVED** [review]
Approving because I managed to generate a video by compiling and running inference using the resulting NEFFs. 

General comments are non-blocking, but if you choose to resolve the comments, please make sure you remove hardcoded paths and check for existence of files when you're loading them.

**Review by jimburtoft — COMMENTED** [review]
Addressed all comments in edf03da:
- Removed `load_with_weights.py` entirely (obsolete dev script — its `shard_weight()` function is now inlined in `shard_weights.py`)
- Switched to `os.environ.setdefault()` in all modeling files
- Removed all hardcoded `/home/ubuntu/` paths — key paths are now required CLI args (`--model-path`, `--compile-dir`, `--backbone-sharded-dir`)
- Added file existence checks (`_check_file_exists()`) before loading model files

Thanks for the review!
